synapse/handlers/federation.py

```diff
@@ -784,15 +784,23 @@ async def _process_received_pdu(
                     resync = True
 
             if resync:
-                await self.store.mark_remote_user_device_cache_as_stale(event.sender)
+                run_in_background(self._resync_device, event.sender)
 
-                # Immediately attempt a resync in the background
-                if self.config.worker_app:
-                    return run_in_background(self._user_device_resync, event.sender)
-                else:
-                    return run_in_background(
-                        self._device_list_updater.user_device_resync, event.sender
-                    )
+    async def _resync_device(self, sender: str):
+        """We have detected that the device list for the given user may be out
+        of sync, so we try and resync them.
+        """
+
+        try:
+            await self.store.mark_remote_user_device_cache_as_stale(sender)
```

Comment:
FWIW this function will cause the next iteration of the resync retry loop (introduced in https://github.com/matrix-org/synapse/pull/7453) to retry this device list, which means that we can have 2 resync happen at the same time for a given user. This may be fine, but at least we should be aware of it.

Reply:
Oh, interesting. I think its still correct to go and fire off a resync, as a) we don't want to wait for the loop to come round necessarily, b) its safe and c) I don't really want to assume that the loop will come round quickly or even retry this particular user (it might have backoff logic etc)

Reply:
Yeah it makes sense, again I think it's fine as long as we're aware of it :)
. o O ( we could avoid this issue by storing and checking an in-memory list of the users we're currently resyncing (i.e. update it in `user_device_resync`) but that's probably out of scope here )